File: 61_rotate_right/rotate_right.rs
/*
 * @Date: 2021-03-27 07:54:39
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-03-27 08:28:41
 */

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

fn rotate_right(mut head: Option<Box<ListNode>>, k: i32) -> Option<Box<ListNode>> {
    debug_assert!(k >= 0);

    if head.is_none() {
        return None;
    }

    if k == 0 {
        return head;
    }

    let mut len = 0;
    let mut ptr = &head;
    while let Some(ref node) = ptr {
        len += 1;
        ptr = &node.next;
    }

    let step = k % len;

    if step == 0 || len == 1 {
        return head;
    }

    let mut ptr = &mut head;
    for _ in 1..(len - step) {
        ptr = &mut ptr.as_mut().unwrap().next;
    }

    // 找到断开点，断开，后面的就是新 head
    let mut new_head = ptr.as_mut().unwrap().next.take();

    // 找到 tail
    let mut tail = &mut new_head;
    while tail.is_some() && tail.as_ref().unwrap().next.is_some() {
        tail = &mut tail.as_mut().unwrap().next;
    }

    // 把原来的 head 接到 tail 后
    tail.as_mut().unwrap().next = head;

    new_head
}

fn main() {
    let head = ListNode {
        val: 1,
        next: Some(Box::new(ListNode {
            val: 2,
            next: Some(Box::new(ListNode {
                val: 3,
                next: Some(Box::new(ListNode {
                    val: 4,
                    next: Some(Box::new(ListNode { val: 5, next: None })),
                })),
            })),
        })),
    };

    let cur = rotate_right(Some(Box::new(head)), 2);
    println!("{:?}", cur);

    let head = ListNode {
        val: 0,
        next: Some(Box::new(ListNode {
            val: 1,
            next: Some(Box::new(ListNode { val: 2, next: None })),
        })),
    };

    let cur = rotate_right(Some(Box::new(head)), 4);
    println!("{:?}", cur);
}

File: cpp/61_rotate_right/rotate_right.cpp
/*
 * @Date: 2021-03-27 07:54:22
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-03-27 08:21:23
 */

#include <iostream>

struct ListNode {
  int val;
  ListNode* next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode* next) : val(x), next(next) {}
};

ListNode* rotate_right(ListNode* head, int k) {
  if (head == nullptr || head->next == nullptr) return head;
  ListNode* last = head;
  int count = 1;

  // 计算链表总长
  while (last->next != nullptr) {
    last = last->next;
    ++count;
  }

  ListNode* front = head;
  // 必要的优化
  if (k > count) k %= count;

  // 链表头拼接到末尾
  while (count - k) {
    last->next = front;
    front = front->next;
    last = last->next;
    ++k;
  }
  last->next = nullptr;
  return front;
}

int main() {
  {
    ListNode* head_1 = new ListNode(
        1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5)))));
    ListNode* cur = rotate_right(head_1, 2);
    while (cur != nullptr) {
      std::cout << cur->val << ", ";
      cur = cur->next;
    }
  }
  std::cout << std::endl;

  {
    ListNode* head_2 = new ListNode(0, new ListNode(1, new ListNode(2)));
    ListNode* cur = rotate_right(head_2, 4);
    while (cur != nullptr) {
      std::cout << cur->val << ", ";
      cur = cur->next;
    }
  }

  return 0;
}